src/text_util.rs

```diff
@@ -45,3 +45,218 @@ pub fn write_indented(
+    }
+    split_lines
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_split_byte_line_to_words() {
+        assert_eq!(split_byte_line_to_words(b""), vec![]);
+        assert_eq!(
+            split_byte_line_to_words(b"foo"),
+            vec![ByteFragment {
+                word: b"foo",
+                whitespace_len: 0,
+                width: 3
+            }],
+        );
+        assert_eq!(
+            split_byte_line_to_words(b"  foo"),
+            vec![
+                ByteFragment {
+                    word: b"",
+                    whitespace_len: 2,
+                    width: 0
+                },
+                ByteFragment {
+                    word: b"foo",
+                    whitespace_len: 0,
+                    width: 3
+                },
+            ],
+        );
+        assert_eq!(
+            split_byte_line_to_words(b"foo  "),
+            vec![ByteFragment {
+                word: b"foo",
+                whitespace_len: 2,
+                width: 3
+            }],
+        );
+        assert_eq!(
+            split_byte_line_to_words(b"a b  foo bar "),
+            vec![
+                ByteFragment {
+                    word: b"a",
+                    whitespace_len: 1,
+                    width: 1
+                },
+                ByteFragment {
+                    word: b"b",
+                    whitespace_len: 2,
+                    width: 1
+                },
+                ByteFragment {
+                    word: b"foo",
+                    whitespace_len: 1,
+                    width: 3,
+                },
+                ByteFragment {
+                    word: b"bar",
+                    whitespace_len: 1,
+                    width: 3,
+                },
+            ],
+        );
+    }
+
+    #[test]
+    fn test_wrap_bytes() {
+        assert_eq!(wrap_bytes(b"foo", 10), [b"foo".as_ref()]);
+        assert_eq!(wrap_bytes(b"foo bar", 10), [b"foo bar".as_ref()]);
+        assert_eq!(
+            wrap_bytes(b"foo bar baz", 10),
+            [b"foo bar".as_ref(), b"baz".as_ref()],
+        );
+
+        // Empty text is represented as [""]
+        assert_eq!(wrap_bytes(b"", 10), [b"".as_ref()]);
+        assert_eq!(wrap_bytes(b" ", 10), [b"".as_ref()]);
+
+        // Whitespace in the middle should be preserved
+        assert_eq!(
+            wrap_bytes(b"foo  bar   baz", 8),
+            [b"foo  bar".as_ref(), b"baz".as_ref()],
+        );
+        assert_eq!(
+            wrap_bytes(b"foo  bar   x", 7),
+            [b"foo".as_ref(), b"bar   x".as_ref()],
+        );
+
+        // Ends with "\n"
+        assert_eq!(wrap_bytes(b"foo\n", 10), [b"foo".as_ref(), b"".as_ref()]);
+        assert_eq!(wrap_bytes(b"foo\n", 3), [b"foo".as_ref(), b"".as_ref()]);
+        assert_eq!(wrap_bytes(b"\n", 10), [b"".as_ref(), b"".as_ref()]);
+
+        // Overflow
+        assert_eq!(wrap_bytes(b"foo x", 2), [b"foo".as_ref(), b"x".as_ref()]);
```

Comment:
For the usecase of wrapping lines for `jj log` in the teminal, it might be nice to at least have an option of hard-wrapping if it's unavoidable. So, `warp_bytes(b"foo x", 2)` would wrap.